List Patterns Refactor (#254)
pxblue/doc-it: src/docs/patterns/lists.mdx

import GlobalActionPanelHeaderImage from '../../app/assets/design-patterns/lists/action-button-in-panel-header.png';
import GlobalActionAsButtonImage from '../../app/assets/design-patterns/lists/action-button-as-button.png';
import SortableListImage from '../../app/assets/design-patterns/lists/sortable-list.gif';
import LocalListActionsImage from '../../app/assets/design-patterns/lists/local-list-actions.png';
import MultiSelectImage from '../../app/assets/design-patterns/lists/multi-select-list.png';
import NavigateListImage from '../../app/assets/design-patterns/lists/navigate-list.png';

<!-- Specs -->

import Spec1Image from '../../app/assets/design-patterns/lists/spec-1.png';
import Spec2Image from '../../app/assets/design-patterns/lists/spec-2.png';
import Spec3Image from '../../app/assets/design-patterns/lists/spec-3.png';
import Spec4Image from '../../app/assets/design-patterns/lists/spec-4.png';
import Spec5Image from '../../app/assets/design-patterns/lists/spec-5.png';
import SpecTypographyImage from '../../app/assets/design-patterns/lists/spec-typography.png';

<!-- KEYWORDS: design pattern list status sort table drag drop action data -->

> # Lists & Tables

Author: Personally, lists and tables are just two sides of one coin. Lists focus more on individual items (you read them by row), while tables focus more on trends (you read them by column). The difference between the two is blurred — how many components can I add to a list before it turns into a table? Does a list with some bare table headers suddenly become a table?

> # Lists & Tables

Reviewer: I agree that the distinction between the two is muddled. 

I don't necessarily agree that a table is _read_ by column or that it is meant to focus on trends — You read a table holistically (if you just look at a column it means nothing without knowing what the row is, and often you're looking for correlations between columns) and if you are interested in trends, you wouldn't try to find them in raw data, you'd use a chart / graph. And I don't think quantity of items is a deciding factor — you can have a list with 10k items and a table with 3 rows and 2 columns. It's more about the structure of the data. 

I do think you're on the right track though...when you're thinking about 'columns', that's when you're getting into table territory and when you're thinking more about the 'item' it's more of a list. 

I am struggling to come up with a definition of when to use one versus the other. It's kind of a 'know it when you see it' kind of thing.

If we're going to discuss tables in this pattern though, then we need to be clear on what the difference is.

My biggest qualm is that List components are not well suited for building tables (because there are no columns), so I would not want to encourage people to use them for 'the same things.'

> # Lists & Tables

Reviewer: Maybe the distinction can be around whether you need column labels or not. If it's not clear what the data is without headers, then it's a table. If you can understand the entire item without labels, it's a list?

> # Lists & Tables

Author: Discussion during the OH — go with the status quo, log an issue for it in the backlog to talk to studio blue about it

<ImageGrid images={[ListBannerImage]} regularWidth />

Lists and tables are frequently used throughout PX Blue applications. Typically, lists are used to represent groups of similar data. They can display from simple name-value pairs in a sidebar all the way to complex timelines or device lists.

<MaterialDesignDescription title={'Lists'} url={'https://material.io/components/lists/'} />

<!--
A TOC like this would be nice, but apparently these links are not working
-   [Common Use Cases](#common-use-cases)
-   [Variations](#variations)
-   [Behaviors](#behaviors)
-   [Design Specifications](#design-specifications)
-   [Developers](#developers)
-->

<Divider />

# Common Use Cases

<ImageGrid
    images={[
        NotificationListDesktopImage,
        NotificationListMobileImage,
        DashboardListDesktopImage,
        DashboardListMobileImage,
    ]}
    caption={[
        'Notifications and status list are the two most common use case for the list pattern.',
        '1&2: Notifications, for mobile and desktop respectively.',
        '3&4: A status list embedded inside a dashboard, for mobile and desktop respectively.',

>     ]}

Author: I can move the mobile pics to the behavior section that talks about the responsiveness

>     ]}

Author: > Do we want to be advertising the Rail variant of the drawer for desktop images?

I won't mind and I think the design is fairly mature right now, even if we don't have them available in all three component libraries.

>     ]}

Reviewer: I suppose the question is, are we pushing the Rail a the go-to navigation solution? In my mind it's third-tier (after permanent and persistent) for applications where space is a concern.

>     ]}

Reviewer: I don't mind showing it, just want to be cautious about where we are steering people.

>     ]}

Author: Replaced a few images using regular drawer

/>

The scenarios in which lists may occur include:

-   Dashboard / current status
-   Notifications
-   Inventories, list of assets
-   Contacts
-   File tree
-   Logs

> -   Settings

Author: It is used -- see my comments below.

Lists should **NOT** be used where:

-   There are no comparable data presents, such as data sets with different structures;

> -   The users demand a summary rather than detailed data value, such as on the home screen. Use [Data visualization](/patterns/visualizations) to display trends, or [heroes](https://pxblue-components.github.io/react/?path=/info/components-hero--get-read-me-story) to display a summary.

Author: I am trying to find a way to say "don't show a bulky device log on the home screen when you should show summaries"

> -   The users demand a summary rather than detailed data value, such as on the home screen. Use [Data visualization](/patterns/visualizations) to display trends, or [heroes](https://pxblue-components.github.io/react/?path=/info/components-hero--get-read-me-story) to display a summary.

Author: And what is a logical item?

> -   The users demand a summary rather than detailed data value, such as on the home screen. Use [Data visualization](/patterns/visualizations) to display trends, or [heroes](https://pxblue-components.github.io/react/?path=/info/components-hero--get-read-me-story) to display a summary.

Reviewer: Yeah, this differentiation needs a little work. A 'logical item' is like something that matches with your mental model of a 'thing' — like a 'device' or an 'alarm' or a 'channel' rather than something more abstract. But rows in a table could also be a logical item like a 'User'.

> -   The users demand a summary rather than detailed data value, such as on the home screen. Use [Data visualization](/patterns/visualizations) to display trends, or [heroes](https://pxblue-components.github.io/react/?path=/info/components-hero--get-read-me-story) to display a summary.

Reviewer: I don't know if the part about bulky lists needs to be said. Maybe your application is very log heavy or notification heavy and the homescreen is a big long list of something. I think that really comes down to the product and the design. I don't know if we can make a blanket statement like "Don't use a long list on a home screen"

Lists usually reside in some larger structures, such as a panel (mobile) or in a card (desktop).

Or if the list is the only element, by itself.

<React.Fragment>
    <ImageGrid
        images={[AnatomyImage]}
        regularWidth
        caption={[
            'An example of a list embedded inside a panel.',
            '1: Panel header and footer.',
            '2: A global list action, in this case it sits inside the panel header.',
            '3: Some panel modules, providing supporting information or additional interactions to the list.',
            '4: A list item.',

>             '5: A local list action.',

Author: ~It is what it is~ I checked and they are the same size.

        'https://stackblitz.com/github/pxblue/angular-design-patterns?file=src%2Fapp%2Fpages%2Flist%2Fdata-list%2Fdata-list.component.ts'
    }
    react={
        'https://codesandbox.io/s/github/pxblue/react-design-patterns/tree/master?file=/src/pages/lists/data-list/index.tsx&initialpath=data-list'
    }
    ionic
    reactNative
/>

The simplest type of list simply presents a single line (or multiple lines) of text for each element in the collection.

## Lists with Status

<ImageGrid images={[StatusListImage]} regularWidth caption={`Use an icon avatar and / or a status stripe to catch people's attention.`}/>
<DemoCard repository={'status-list'} angular={'https://stackblitz.com/github/pxblue/angular-design-patterns?file=src%2Fapp%2Fpages%2Flist%2Fstatus-list%2Fstatus-list.component.ts'} react={'https://codesandbox.io/s/github/pxblue/react-design-patterns/tree/master?file=/src/pages/lists/status-list/index.tsx&initialpath=status-list'} ionic reactNative/>

Commonly in PX Blue applications, list items also show some visual indication of the item's current status, such as online, offline, or an error condition. This is achieved by using avatars and status stripes.

<ImageGrid

>     images={[IconAlignImage]}

Author: > maybe a space measurement indicator to show the spacing from the text? 

You mean the kind of indicator we see in the specs section?

>     images={[IconAlignImage]}

Reviewer: yeah, just a thought. Moving the dotted line over is probably enough.

>     images={[IconAlignImage]}

Author: Moved the dotted line. I like the way it looks right now — I think all the specs should be put together in one section, rather than scattered everywhere.

    regularWidth
    caption={
        'Whenever you have a list that may include both avatars and icons, make sure all the icons, with or without avatars, are center-aligned. If you are using interactive elements like checkboxes and radio buttons, they should also be center-aligned to accomodate ripples. Otherwise, all icons should be left-aligned.'
    }
/>

In the rare case where you need to display status but cannot fit in neither an avatar nor a status stripe, such as the mini-list inside a [scorecard component](https://pxblue-components.github.io/react/?path=/info/components-score-card--get-read-me-story), you may color-code your list item icons and labels. However, for accessibility concerns, you should never rely _solely_ on the color to convey the message.

<ImageGrid
    images={[ScorecardImage]}
    regularWidth
    caption={[
        'A scorecard with a mini-list in the card body.',
        `1: In addition to the change of color for the icon, you can also change icons, make label text descriptive, and bolden the font.`,
    ]}
/>

Using tags is another way to display list status. They typically appear at the very end of a list item.

> <ImageGrid

Author: > Is this something we would actually expect someone to do in an application? 

It is stolen and sanitized from GP. Guessing from their design file, I think the status stripe is meant to highlight anything attention worthy (like your own account, people pending approves)

> <ImageGrid

Reviewer: It's not my favorite design, but it can slide. I would maybe split them out and show three separate examples of how a list item could indicate status (one for stripe, one for badge, one for LIT), but I like that this image has some context and shows them all at once.

> <ImageGrid

Author: I took out the status stripe, as it is already covered by previous examples. I think the remaining ones, dotted badge works fine, with the tag here

    images={[ContactsImage]}
    regularWidth
    caption={'This user list uses three status indicators: a status stripe, dot badges, and tags.'}
/>

## Timestamps

> Many UI elements can be added to the list items, and one of the most common such elements is a timestamp, which usually occurs in device logs and list of notifications. You are encouraged to group your list items by dates for readability.

Author: Depends on the actual usage (imagine a device log list with only three logs after 5 years of usage), I won't go as harsh as "you _should_ group your list items by date".

> Many UI elements can be added to the list items, and one of the most common such elements is a timestamp, which usually occurs in device logs and list of notifications. You are encouraged to group your list items by dates for readability.

Author: Changed back to "you are encouraged to group your list items by date for readability."

> Many UI elements can be added to the list items, and one of the most common such elements is a timestamp, which usually occurs in device logs and list of notifications. You are encouraged to group your list items by dates for readability.

Reviewer: Well that's where some application logic comes into play...we don't dictate what the buckets are, just that you should have buckets. If you have three logs after 5 years, you'd have one bucket: "Last 5 years".

> Many UI elements can be added to the list items, and one of the most common such elements is a timestamp, which usually occurs in device logs and list of notifications. You are encouraged to group your list items by dates for readability.

Reviewer: Or "If you have a large number of items in the list, you should group them by date to make them easier to scan"

<ImageGrid
    images={[NotificationBucketsImage1, NotificationBucketsImage2]}
    caption={[
        'You may organically group your lists into buckets.',
        '1: "Today - Yesterday - Earlier This Week" buckets',
        '2: Plain date buckets. Note the dates have been taken out of the individual list items to avoid redundancy.',
    ]}
/>

> ## Tables

Author: Per discussion we'd sit on this.

<React.Fragment>
    <Divider />
</React.Fragment>

# Behaviors

Lists can be interactive. In general, there are two types of interactions: global list actions and local list actions.

<DemoCard
    repository={'action-list'}
    angular={
        'https://stackblitz.com/github/pxblue/angular-design-patterns?file=src%2Fapp%2Fpages%2Flist%2Faction-list%2Faction-list.component.ts'
    }
    react={
        'https://codesandbox.io/s/github/pxblue/react-design-patterns/tree/master?file=/src/pages/lists/action-list/index.tsx&initialpath=action-list'
    }
    ionic
    reactNative
/>

## Global List Actions

Global list actions are the actions that manipulate the entire list, rather than treating each list item individually. The buttons typically sit in the [Appbar](/patterns/appbar), the panel header, or a button row (desktop-only).

<ImageGrid
    images={[GlobalActionAppBarImage, GlobalActionPanelHeaderImage, GlobalActionAsButtonImage]}
    caption={[
        `1: Global actions in the Appbar.`,
        '2: Global actions in the panel header.',

>         '3: Global actions in its own button row.',

Reviewer: Button panel works. I'm not a huge fan of these 'floating' button panels, but I won't fight you on it.

    ]}
/>

Common global list actions include:

-   Adding a new list item
-   Sorting and filtering the list
-   Toggling a different view, such as a grid view
-   Exporting or downloading the list data

### Drag and Drop

<ImageGrid images={[SortableListImage]} regularWidth />
<DemoCard repository={'sortable-list'} angular={'https://stackblitz.com/github/pxblue/angular-design-patterns?file=src%2Fapp%2Fpages%2Flist%2Fsortable-list%2Fsortable-list.component.ts'} react={'https://codesandbox.io/s/github/pxblue/react-design-patterns/tree/master?file=/src/pages/lists/sortable-list/index.tsx&initialpath=sortable-list'} ionic reactNative />

> Drag and drop is a special case for list sorting. It re-orders the items in a list manually. This should be accomplished by placing a drag handle as the leftmost element in the row. The ability to re-order the list should be locked behind an Edit Mode toggle mechanism.

Author: I think everything should be pushed to the side, including the icons. 

![ezgif-6-9b2b20979da3](https://user-images.githubusercontent.com/8997218/93805415-b9351d00-fc15-11ea-938b-a95968d5cc31.gif)

This just looks awkward to me:
![ezgif-6-cb0e8e74f270](https://user-images.githubusercontent.com/8997218/93805416-b9cdb380-fc15-11ea-9340-74350bb42e19.gif)

> Drag and drop is a special case for list sorting. It re-orders the items in a list manually. This should be accomplished by placing a drag handle as the leftmost element in the row. The ability to re-order the list should be locked behind an Edit Mode toggle mechanism.

Author: I can't think of a use case when we don't lock the list reordering behind an edit mode

> Drag and drop is a special case for list sorting. It re-orders the items in a list manually. This should be accomplished by placing a drag handle as the leftmost element in the row. The ability to re-order the list should be locked behind an Edit Mode toggle mechanism.

Reviewer: Unfortunately, our ILI doesn't support multiple icons like that 🤦 . 
"Don't use icons if your list is sortable!"

> Drag and drop is a special case for list sorting. It re-orders the items in a list manually. This should be accomplished by placing a drag handle as the leftmost element in the row. The ability to re-order the list should be locked behind an Edit Mode toggle mechanism.

Reviewer: File tree? You can leave it as you have it though, that's fine.

> Drag and drop is a special case for list sorting. It re-orders the items in a list manually. This should be accomplished by placing a drag handle as the leftmost element in the row. The ability to re-order the list should be locked behind an Edit Mode toggle mechanism.

Author: LOL! Time for some rightComponents hack

## Local List Actions

<ImageGrid
    images={[LocalListActionsImage]}
    caption={[

>         '1: A more_vert icon, used to open a menu for more list actions.',

Author: I am happy with 3-dot

Material Design prefers `more_horiz` for iOS platforms, but we've been using more_vert everywhere. So "3-dot" might not be a confusing name.

>         '2: A chevron, usually opens up a new page with list item details.',

Reviewer: I think normally the chevron is just an icon and the entire row is clickable to navigate to the new page. I don't think we would typically see a list-item that has actions and a chevron (like number 4).

We may want to flesh this out a bit.

Author: good point. I will take out #4 and treat chevron a "special icon" that indicates the navigation you mentioned above.

Author: Changed the caption to 

"2: A chevron, usually indicates the entire list item is clickable and opens up a new page with list item details."

        '3: A few icons that serve as list action buttons.',
        '4: List actions can be used together.',
        '5: An inline checkbox to toggle a status.',

>         '6: An inline text field for data entry.',

Reviewer: This one looks a little more like a table than a ListItem.

Author: Per discussion: we are gonna sit on this

    ]}
    gridImageProps={{ sm: 8 }}
/>

> If there are actions that you would like to perform on individual list items, you may add interactive elements such as icons or a 3-dot action menu. Depending on the space available, you typically will want to use the action menu if you have more than three actions that can be taken.

Reviewer: ```suggestion
If there are actions that you need to perform on individual list items, you may add icon buttons or an action menu. If you have three or fewer actions, they can be represented by individual icon buttons. If there are more than three, you should move them into an action menu.
```

> > If your list item is clickable for the entire item, then you should not arrange any extra local list actions, as the two interactions can intervene with each other.

Reviewer: ```suggestion
> If your entire list item is clickable, you should not include local list actions, as the two interactions can interfere with one another.
```

Author: 20 years after I first started learning English, I still can't tell the precise difference between intervene and interfere... 😞

> ### Navigating Down a Tree Structure

Reviewer: Should this come after the actions on selected items section to keep all the 'actions' stuff together?

Author: This counts as a local list item action, so I prefer to leave this here.

Reviewer: I can see it either way. I was sort of seeing the progression as Actions on a Single item -> Actions on Multiple items -> Special Cases. But if you're considering it to be a special case of action on a single item, it makes sense.

Should be bump up the indentations on the file tree to make the levels more clear?

<ImageGrid
    images={[NavigateListImage]}
    caption={['1: A list used for separated pages.', '2: A list displaying a folder structure.']}
    regularWidth

> />

Reviewer: I find the second image difficult to read. Is this something that we are encouraging teams to do? Is this common?

Author: Is that a problem with the divider? Is this slightly better? Or is it about indentations?

![image](https://user-images.githubusercontent.com/8997218/93806675-83913380-fc17-11ea-8720-d15dbf509a89.png)

The example is retrieved and sanitized from PredictPulse.

Author: I think PredictPulse is the only project I can find that involves a file tree like this.

Reviewer: I think it was more of an indentation thing. If I squint my eyes everything looks like it's the same level.

Author: Bumped the indentation up. Now it not only has the same indentation as our drawer component, the folder icon also lines up with its parent folder's text. Hooray!

Lists can be used to navigate a tree-like structure in an application. See [the navigation pattern](/patterns/navigation).

### Actions on Selected Items

> <ImageGrid images={[MultiSelectImage]} regularWidth />

Reviewer: This image seems like a table more than a list

<DemoCard repository={'multiselect-list'} angular={'https://stackblitz.com/github/pxblue/angular-design-patterns?file=src%2Fapp%2Fpages%2Flist%2Fmultiselect-list%2Fmultiselect-list.component.ts'} react={'https://codesandbox.io/s/github/pxblue/react-design-patterns/tree/master?file=/src/pages/lists/multiselect-list/index.tsx&initialpath=multiselect-list'} ionic reactNative />

> You may want to take action on several list items at once, but not all. The easiest way to do this is through a multi-select list. You may optionally add a highlight color for the selected items.

Reviewer: Can we add more content about where to present the actions? The image uses a button panel, but our example uses a snackbar. Do we have recommendations/guidance around this?

Author: Sounds like it can be a topic for overlay pattern overhaul :P

All the desktop examples I've seen use a button panel, and all the mobile examples use icon buttons in the app bar.

Reviewer: It only goes in overlay if the recommendation is to use overlays 😛

Author: Or to not use overlays? 😝

## Responsive Design

> When it comes to responsive design and progressive disclosure, it can be tricky to fit all the information from a list onto a smaller screen. Here are some best practices on designing for mobile phone:

Reviewer: ```suggestion
When it comes to responsive design and progressive disclosure, it can be tricky to fit all the information from a list onto a smaller screen. Here are some best practices on designing for mobile:
```

> -   If a list item has too many elements in it and doesn't fit, try converting the item into a summary card, with a "view more" option to view the full data.
> -   Avoid using tables.
> -   Try adding gestures (long-press, swipe-left) that fits with the phone's OS ecosystem.

Reviewer: ```suggestion
-   Avoid using tables if possible, or reduce the number of columns visible on mobile
-   Remove less critical information from lists on mobile
-   Use summary cards with a "view more" option to view the full data.
-  Use gestures (e.g., long-press, swipe-left) that align with the target device's behavior
```

<Divider />

# Design Specifications

> <ImageGrid images={[Spec1Image, Spec2Image, Spec3Image, Spec4Image, Spec5Image, SpecTypographyImage]} regularWidth />

Reviewer: Specs 1:
Text with Badges? --> Text with Action buttons?

Specs 2:
should we specify that the icon is 24?
Should the spec present the avatar space as 40, the way it's implemented (as you have it), or should we present it as the size of the icon/avatar and the right margin (the way it looks)?

Specs 4:
I don't think the timestamp section is hardcoded with a width. It's just a 32 (i think) margin between leftContent and the title.

Author: > Should the spec present the avatar space as 40, the way it's implemented (as you have it), or should we present it as the size of the icon/avatar and the right margin (the way it looks)?

I personally prefer the way it is implemented, bc it puts more rationales for the anatomy

Author: > I don't think the timestamp section is hardcoded with a width. It's just a 32 (i think) margin between leftContent and the title.

If it is a margin, then your title & subtitle will shift left & right depends on the timestamp length.

Reviewer: In the component it's implemented as a margin, so unless you go in and set a width yourself, then the other content will shift, yes. We can't set a width in the component because the left content can be _anything_, not just a timestamp.

If we are saying 'in the specific case of a timestamp, you should make the timestamp 80dp wide',  I'm fine with that.

Author: yeah that is what I'm trying to do here

<MaterialDesignDescription title={'Lists'} description={`For additional instructions, refer to Material Design's design specs`} url={'https://material.io/components/lists#specs'} />
<MaterialDesignDescription title={'Data Tables'} description={`For additional instructions, refer to Material Design's design specs`} url={'https://material.io/components/data-tables#specs'} />
<Divider />

# Developers

Use the following components to implement this pattern:

**Angular** and **Ionic**:

-   Angular Material
    -   [List](https://material.angular.io/components/list/overview)
    -   [Table](https://material.angular.io/components/table/overview)
-   @pxblue/angular-components
    -   [Info List Item](https://pxblue-components.github.io/angular/?path=/info/components-info-list-item--readme)
    -   [List Item Tag](https://pxblue-components.github.io/angular/?path=/info/components-list-item-tag--readme)

> In Angular, the CDK (Component Developer's Kit) greater than 7.0 is required for the drag-and-drop feature.

Reviewer: ```suggestion
> In Angular, the CDK (Component Developer's Kit) version 7+ is required to use the drag-and-drop feature.
```

**React**:

-   Material UI
    -   [List](https://material-ui.com/api/list/)
    -   [List Item](https://material-ui.com/api/list-item/)
-   @pxblue/angular-components
    -   [Info List Item](https://pxblue-components.github.io/react/?path=/info/components-info-list-item--get-read-me-story)
    -   [List Item Tag](https://pxblue-components.github.io/react/?path=/info/components-list-item-tag--get-read-me-story)

**React Native**:

-   React-Native
    -   [Flat List](https://reactnative.dev/docs/using-a-listview)

> -   React Native Elements

Reviewer: Take out React Native elements...it's not our recommended component library anymore (just RNP now)

Author: Ok. I got the idea from our React Native component library showcase. Should I log an issue in that repo?

Reviewer: I believe we have a story in JIRA to overhaul that showcase and make it a shared submodule (R18). If you want to check the criteria for that and make sure it says to remove RNE, that should suffice. RNE was previously our preferred component library, but RNP has pulled ahead and is more material-y.
